tls/s2n_handshake_io.c

```diff
@@ -852,6 +852,20 @@ message_type_t s2n_conn_get_current_message_type(struct s2n_connection *conn)
     return ACTIVE_MESSAGE(conn);
 }
 
+static S2N_RESULT s2n_execute_handler(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* Ensure the state machine referenced is consistant throughout the handshake */
+    if (conn->actual_protocol_version_established) {
```

Comment:
We might need it to be a little more complicated than this. The actual protocol is established after the ServerHello, but we can start setting TLS1.3-only flags / committing to TLS1.3 earlier due to early data: https://github.com/aws/s2n-tls/blob/main/tls/extensions/s2n_client_early_data_indication.c#L25-L49

You should also probably investigate / test how actual_protocol_version_established interacts with retries.

Reply:
Maybe set something the first time we set a tls1.2 / tls1.3 flag? https://github.com/aws/s2n-tls/blob/main/tls/extensions/s2n_client_early_data_indication.c#L25-L49t

Reply:
The issue with s2n_handshake_type_set_tls13_flag() is that is isn't called in all valid TLS13 handshakes. Like a handshake that is just [NEGOTIATED | FULL_HANDSHAKE] doesn't ever set any TLS13 flags, but it is still a valid TLS13 handshake. I could set conn->is_tls13_state_machine there, but it would have to be in addition to where it's set now.

Reply:
I could set something in s2n_conn_set_tls13_handshake_type() as that is called after the server reads the client hello, which is slightly earlier than this PR.

Reply:
I don't think just "earlier" is enough. It should be when we set the TLS1.3-only flag. But yeah, setting it either when we set a TLS1.3 flag or where you're currently setting it sounds reasonable.

Reply:
I added it to s2n_handshake_type_set_tls13_flag(). This gets called for both a HRR handshake and an Early Data handshake.